routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,473 @@
+	if err != nil {
+		log.Error("Error parsing upload URL: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	log.Debug("[artifact] get upload url: %s, artifact id: %d", url, artifact.ID)
+
+	ctx.JSON(http.StatusOK, map[string]interface{}{
+		"fileContainerResourceUrl": url,
+	})
+}
+
+// getUploadFileSize returns the size of the file to be uploaded.
+// The raw size is the size of the file as reported by the header X-TFS-FileLength.
+func (ar artifactRoutes) getUploadFileSize(ctx *context.Context) (int64, int64, error) {
+	contentLength := ctx.Req.ContentLength
+	xTfsLength, _ := strconv.ParseInt(ctx.Req.Header.Get(artifactXTfsFileLengthHeader), 10, 64)
+	if xTfsLength > 0 {
+		return xTfsLength, contentLength, nil
+	}
+	return contentLength, contentLength, nil
+}
+
+func (ar artifactRoutes) saveUploadChunk(ctx *context.Context,
+	artifact *actions.ActionArtifact,
+	contentSize, taskID int64,
+) (int64, error) {
+	// read chunk data to calc md5
+	chunkData, err := io.ReadAll(ctx.Req.Body)
+	if err != nil {
+		return -1, err
+	}
+
+	// check md5
+	md5Hash := md5.Sum(chunkData)
+	md5String := base64.StdEncoding.EncodeToString(md5Hash[:])
+	if md5String != ctx.Req.Header.Get(artifactXActionsResultsMD5Header) {
+		return -1, fmt.Errorf("md5 not match")
+	}
+
+	contentRange := ctx.Req.Header.Get("Content-Range")
+	start, end, length := int64(0), int64(0), int64(0)
+	if _, err := fmt.Sscanf(contentRange, "bytes %d-%d/%d", &start, &end, &length); err != nil {
+		return -1, fmt.Errorf("parse content range error: %v", err)
+	}
+
+	storagePath := fmt.Sprintf("tmp%d/%d-%d-%d.chunk", taskID, artifact.ID, start, end)
+
+	// save chunk to storage
+	if _, err := ar.fs.Save(storagePath, bytes.NewBuffer(chunkData), -1); err != nil {
+		return -1, fmt.Errorf("save chunk to storage error: %v", err)
+	}
+
+	log.Debug("[artifact] save chunk %s, size: %d, artifact id: %d, start: %d, end: %d",
+		storagePath, contentSize, artifact.ID, start, end)
+
+	return length, nil
+}
+
+var invalidArtifactNameChars = []string{"\\", "/", "\"", ":", "<", ">", "|", "*", "?", "\r", "\n"}
+
+// checkArtifactName checks if the artifact name contains invalid characters.
+// If the name contains invalid characters, an error is returned.
+// The rules are from https://github.com/actions/toolkit/blob/main/packages/artifact/src/internal/path-and-artifact-name-validation.ts#L32
+func checkArtifactName(name string) error {
+	for _, c := range invalidArtifactNameChars {
+		if strings.Contains(name, c) {
+			return fmt.Errorf("artifact name contains invalid character %s", c)
+		}
+	}
+	return nil
+}
+
+func (ar artifactRoutes) uploadArtifact(ctx *context.Context) {
+	artifactID := ctx.ParamsInt64("artifactID")
+
+	artifact, err := actions.GetArtifactByID(ctx, artifactID)
+	if err != nil {
+		log.Error("Error getting artifact: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	itemPath := ctx.Req.URL.Query().Get("itemPath")
```

Comment:
I guess it could wait for #23495 (path clean helpers).

Otherwise the question always comes: is it safe to split it or use it directly?